File: library/src/main/java/com/trinity/encoder/MediaCodecSurfaceEncoder.java
package com.trinity.encoder;
import android.annotation.SuppressLint;
import android.annotation.TargetApi;
import android.media.MediaCodec;
import android.media.MediaCodec.BufferInfo;
import android.media.MediaCodecInfo;
import android.media.MediaFormat;
import android.media.MediaMuxer;
import android.os.Build;
import android.os.Bundle;
import android.util.Log;
import android.view.Surface;

import java.io.IOException;
import java.nio.ByteBuffer;

@TargetApi(Build.VERSION_CODES.JELLY_BEAN_MR2)
public class MediaCodecSurfaceEncoder {
    private static final String TAG = "MediaCodecEncoder";
    private static final boolean VERBOSE = false;
    private static final int MEDIA_CODEC_NOSIE_DELTA = 80*1024;

    public static final String MIME_TYPE = "video/avc"; // H.264 Advanced Video
    // Coding
    private static final int IFRAME_INTERVAL = 1; // sync frame every second

    private Surface mInputSurface;
    private MediaCodec mEncoder;
    private BufferInfo mBufferInfo;
    private long lastPresentationTimeUs = -1;

    private MediaMuxer mMuxer;
    private int mTrackIndex;

    public MediaCodecSurfaceEncoder() {

    }

    public MediaCodecSurfaceEncoder(int width, int height, int bitRate, int frameRate, String outputPath) throws Exception {
    	Log.i("problem", String.format("MediaCodecSurfaceEncoder width:%d, height:%d", width, height));
        MediaFormat format = MediaFormat.createVideoFormat(MIME_TYPE, width, height);
        format.setInteger(MediaFormat.KEY_COLOR_FORMAT, MediaCodecInfo.CodecCapabilities.COLOR_FormatSurface);
        format.setInteger(MediaFormat.KEY_BIT_RATE, bitRate);
        format.setInteger(MediaFormat.KEY_FRAME_RATE, frameRate);
        format.setInteger(MediaFormat.KEY_I_FRAME_INTERVAL, IFRAME_INTERVAL);
        if (VERBOSE)
            Log.d(TAG, "format: " + format);
        try {
            mEncoder = MediaCodec.createEncoderByType(MIME_TYPE);
            mEncoder.configure(format, null, null, MediaCodec.CONFIGURE_FLAG_ENCODE);
            mInputSurface = mEncoder.createInputSurface();
            mEncoder.start();
        } catch (Exception e) {
            throw e;
        }

        try {
            mMuxer = new MediaMuxer(outputPath, MediaMuxer.OutputFormat.MUXER_OUTPUT_MPEG_4);
        } catch (IOException ioe) {
            throw new RuntimeException("MediaMuxer creation failed", ioe);
        }
        mTrackIndex = -1;
    }

    public MediaCodecSurfaceEncoder(int width, int height, int bitRate, int frameRate) throws Exception {
    	Log.i("problem", String.format("MediaCodecSurfaceEncoder width:%d, height:%d", width, height));
        MediaFormat format = MediaFormat.createVideoFormat(MIME_TYPE, width, height);
        format.setInteger(MediaFormat.KEY_COLOR_FORMAT, MediaCodecInfo.CodecCapabilities.COLOR_FormatSurface);
        format.setInteger(MediaFormat.KEY_BIT_RATE, bitRate);
        format.setInteger(MediaFormat.KEY_FRAME_RATE, frameRate);
        format.setInteger(MediaFormat.KEY_I_FRAME_INTERVAL, IFRAME_INTERVAL);
        if (VERBOSE)
            Log.d(TAG, "format: " + format);
        try {
            mEncoder = MediaCodec.createEncoderByType(MIME_TYPE);
            mEncoder.configure(format, null, null, MediaCodec.CONFIGURE_FLAG_ENCODE);
            mInputSurface = mEncoder.createInputSurface();
            mEncoder.start();
        } catch (Exception e) {
            throw e;
        }

        mTrackIndex = -1;
    }

    @SuppressLint("NewApi")
	public void hotConfig(int width, int height, int bitRate, int frameRate) throws Exception {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            try {
                if (mEncoder != null) {
                    mEncoder.reset();
                }

                MediaFormat format = MediaFormat.createVideoFormat(MIME_TYPE, width, height);
                format.setInteger(MediaFormat.KEY_COLOR_FORMAT, MediaCodecInfo.CodecCapabilities.COLOR_FormatSurface);
                format.setInteger(MediaFormat.KEY_BIT_RATE, bitRate - MEDIA_CODEC_NOSIE_DELTA);
                format.setInteger(MediaFormat.KEY_FRAME_RATE, frameRate);
                format.setInteger(MediaFormat.KEY_CAPTURE_RATE, frameRate);
                format.setInteger(MediaFormat.KEY_I_FRAME_INTERVAL, IFRAME_INTERVAL);
                mEncoder.configure(format, null, null, MediaCodec.CONFIGURE_FLAG_ENCODE);
                if(null != mInputSurface) {
                    mInputSurface.release();
                }
                mInputSurface = mEncoder.createInputSurface();
                mEncoder.start();

                Log.i(TAG, String.format("HotConfig success bitRate:%d, frameRate:%d", bitRate, frameRate));
            } catch (Exception e) {
                throw e;
            }
        }
    }

    interface MuxerCallback {
        void startMux(MediaFormat mediaFormat);

        void writePacket(ByteBuffer encodedData, BufferInfo bufferInfo);

        void stopMux();
    }

    private MuxerCallback muxerCallback = new MuxerCallback() {
        @Override
        public void startMux(MediaFormat mediaFormat) {
            mTrackIndex = mMuxer.addTrack(mediaFormat);
            mMuxer.start();
        }

        @Override
        public void writePacket(ByteBuffer encodedData, BufferInfo bufferInfo) {
            mMuxer.writeSampleData(mTrackIndex, encodedData, bufferInfo);
        }

        @Override
        public void stopMux() {
            try {
                if (-1 != mTrackIndex && mMuxer != null) {
                    mMuxer.stop();
                    mMuxer.release();
                    mMuxer = null;
                }
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    };

    /**
     * Returns the encoder_'s input surface.
     */
    public Surface getInputSurface() {
        return mInputSurface;
    }

    /**
     * Shuts down the encoder_ thread, and releases encoder_ resources.
     * <p>
     * Does not return until the encoder_ thread has stopped.
     */
    public void shutdown() {
        if (VERBOSE)
            Log.d(TAG, "releasing encoder_ objects");

        try {
            if (null != mEncoder) {
                mEncoder.stop();
                mEncoder.release();
                mEncoder = null;
            }
        } catch (Exception e) {
            e.printStackTrace();
        }

        try {
            if (null != muxerCallback) {
                muxerCallback.stopMux();
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private static final int TIMEOUT_USEC = 5000;

    //4.4系统才有的方法
    @TargetApi(Build.VERSION_CODES.KITKAT)
    public void configureBitRate(int targetBitRate) {
        Bundle bitRateBundle = new Bundle();
        bitRateBundle.putInt(MediaCodec.PARAMETER_KEY_VIDEO_BITRATE, targetBitRate);
        mEncoder.setParameters(bitRateBundle);
    }

    //TODO:下面这种方式暂时不使用，由于某一些机器MediaCodec的Bug，而是在底层自己destroy然后在重新创建的模式实现ReConfig
    public void reConfigureFromNative(int targetBitRate) {
        Log.i("problem", "reConfigureFromNative : " + targetBitRate);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
            configureBitRate(targetBitRate);
        }
    }

    /**
     * Drains all pending output from the decoder, and adds it to the circular
     * buffer.
     *
     */
    public void drainEncoder() {
        mBufferInfo = new BufferInfo();
        ByteBuffer[] encoderOutputBuffers = mEncoder.getOutputBuffers();
        int encoderStatus = mEncoder.dequeueOutputBuffer(mBufferInfo, TIMEOUT_USEC);
        if (encoderStatus == MediaCodec.INFO_TRY_AGAIN_LATER) {
            // no output available yet
            Log.i(TAG, "no output available yet");
        } else if (encoderStatus == MediaCodec.INFO_OUTPUT_BUFFERS_CHANGED) {
            Log.i(TAG, " not expected for an encoder_");
            encoderOutputBuffers = mEncoder.getOutputBuffers();
        } else if (encoderStatus == MediaCodec.INFO_OUTPUT_FORMAT_CHANGED) {
            MediaFormat mEncodedFormat = mEncoder.getOutputFormat();
            muxerCallback.startMux(mEncodedFormat);
            Log.d(TAG, "encoder_ output format changed: " + mEncodedFormat);
        } else if (encoderStatus < 0) {
            // let's ignore it
            Log.w(TAG, "unexpected result from encoder_.dequeueOutputBuffer: " + encoderStatus);
        } else {
            ByteBuffer encodedData = encoderOutputBuffers[encoderStatus];
            if (encodedData == null) {
                throw new RuntimeException("encoderOutputBuffer " + encoderStatus + " was null");
            }
            if ((mBufferInfo.flags & MediaCodec.BUFFER_FLAG_CODEC_CONFIG) != 0) {
                // The codec config data was pulled out when we got the
                // INFO_OUTPUT_FORMAT_CHANGED status. The MediaMuxer won't
                // accept
                // a single big blob -- it wants separate csd-0/csd-1 chunks --
                // so simply saving this off won't work.
                if (VERBOSE)
                    Log.d(TAG, "ignoring BUFFER_FLAG_CODEC_CONFIG");
                mBufferInfo.size = 0;
            }
            if (mBufferInfo.presentationTimeUs >= lastPresentationTimeUs) {
                if (mBufferInfo.size != 0) {
                    encodedData.position(mBufferInfo.offset);
                    encodedData.limit(mBufferInfo.offset + mBufferInfo.size);
                    muxerCallback.writePacket(encodedData, mBufferInfo);
                    lastPresentationTimeUs = mBufferInfo.presentationTimeUs;
                    if (VERBOSE) {
                        Log.d(TAG,
                                "sent " + mBufferInfo.size + " bytes to muxer, ts=" + mBufferInfo.presentationTimeUs);
                    }
                } else {
                    Log.i(TAG, "why mBufferInfo.size is equals 0");
                }
            }
            mEncoder.releaseOutputBuffer(encoderStatus, false);

            if ((mBufferInfo.flags & MediaCodec.BUFFER_FLAG_END_OF_STREAM) != 0) {
                Log.w(TAG, "reached end of stream unexpectedly");
            }
        }
    }

    public long pullH264StreamFromDrainEncoderFromNative(byte[] returnedData) {
        long val = 0;

        try {
            mBufferInfo = new BufferInfo();
            ByteBuffer[] encoderOutputBuffers = mEncoder.getOutputBuffers();
            int encoderStatus = mEncoder.dequeueOutputBuffer(mBufferInfo, TIMEOUT_USEC);
            if (encoderStatus == MediaCodec.INFO_TRY_AGAIN_LATER) {
                // no output available yet
                Log.i(TAG, "no output available yet");
            } else if (encoderStatus == MediaCodec.INFO_OUTPUT_BUFFERS_CHANGED) {
                Log.i(TAG, " not expected for an encoder_");
                encoderOutputBuffers = mEncoder.getOutputBuffers();
            } else if (encoderStatus == MediaCodec.INFO_OUTPUT_FORMAT_CHANGED) {
                MediaFormat mEncodedFormat = mEncoder.getOutputFormat();
                Log.d(TAG, "encoder_ output format changed: " + mEncodedFormat);
            } else if (encoderStatus < 0) {
                // let's ignore it
                Log.w(TAG, "unexpected result from encoder_.dequeueOutputBuffer: " + encoderStatus);
            } else {
                ByteBuffer encodedData = encoderOutputBuffers[encoderStatus];
                if (encodedData == null) {
                    throw new RuntimeException("encoderOutputBuffer " + encoderStatus + " was null");
                }
                if ((mBufferInfo.flags & MediaCodec.BUFFER_FLAG_CODEC_CONFIG) != 0) {
                    // The codec config data was pulled out when we got the
                    // INFO_OUTPUT_FORMAT_CHANGED status. The MediaMuxer won't
                    // accept
                    // a single big blob -- it wants separate csd-0/csd-1 chunks --
                    // so simply saving this off won't work.
                    if (VERBOSE)
                        Log.d(TAG, "ignoring BUFFER_FLAG_CODEC_CONFIG, mBufferInfo.size = " + mBufferInfo.size);

                    if (mBufferInfo.size != 0) {
                        encodedData.position(mBufferInfo.offset);
                        encodedData.limit(mBufferInfo.offset + mBufferInfo.size);

                        val = mBufferInfo.size;

                        encodedData.get(returnedData, 0, mBufferInfo.size);
                    } else {
                        Log.i(TAG, "why mBufferInfo.size is equals 0");
                    }

                    mBufferInfo.size = 0;
                }
                if (mBufferInfo.presentationTimeUs >= lastPresentationTimeUs) {
                    if (mBufferInfo.size != 0) {
                        encodedData.position(mBufferInfo.offset);
                        encodedData.limit(mBufferInfo.offset + mBufferInfo.size);
                        lastPresentationTimeUs = mBufferInfo.presentationTimeUs;
                        val = mBufferInfo.size;

                        encodedData.get(returnedData, 0, mBufferInfo.size);

                        int type = returnedData[4] & 0x1F;
                        if(type == 7) {
                            //某些手机I帧前面会附带上SPS和PPS，剥离掉SPS和PPS，仅使用I帧
                            returnedData = H264ParseUtil.splitIDRFrame(returnedData);
                            val = returnedData.length;
                        }

                        if (VERBOSE) {
                            Log.d(TAG,
                                    "sent " + mBufferInfo.size + " bytes to muxer, ts=" + mBufferInfo.presentationTimeUs);
                        }
                    } else {
                        Log.i(TAG, "why mBufferInfo.size is equals 0");
                    }
                } else {
                    Log.d(TAG, "mBufferInfo.presentationTimeUs < lastPresentationTimeUs");
                }

                mEncoder.releaseOutputBuffer(encoderStatus, false);

                if ((mBufferInfo.flags & MediaCodec.BUFFER_FLAG_END_OF_STREAM) != 0) {
                    Log.w(TAG, "reached end of stream unexpectedly");
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        }

        return val;
    }

    public long getLastPresentationTimeUs() {
        return lastPresentationTimeUs;
    }

    // temporary not supported devices list
    public static boolean IsInNotSupportedList() {
//		if(!HWEncoderServerBlackListHelper.isHWEncoderAvailable){
//			return true;
//		}
//		String manufacturer = android.os.Build.MANUFACTURER;
        String model = Build.MODEL;
        if (model.compareTo("Coolpad 8720L") == 0)
            return true;
        if (model.compareTo("vivo X5L") == 0)
            return true;
        if (model.compareTo("CHE-TL00H") == 0)
            return true;
        if (model.indexOf("GT-I9158V") != -1)
            return true;
        if (model.indexOf("HTC D826w") != -1)
            return true;
        if (model.indexOf("y923") != -1)
            return true;
        if (model.indexOf("R7007") != -1)
            return true;
        if (model.indexOf("P6-C00") != -1)
            return true;
        if (model.indexOf("F240L") != -1)    // LG-F240L
            return true;
        if (model.indexOf("A7600") != -1)    // 联想A7600
            return true;

        return false;
    }
}

File: library/src/main/java/com/trinity/encoder/MediaCodecSurfaceEncoder.kt
/*
 * Copyright (C) 2020 Trinity. All rights reserved.
 * Copyright (C) 2020 Wang LianJie <wlanjie888@gmail.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.trinity.encoder

import android.media.MediaCodec
import android.media.MediaCodecInfo
import android.media.MediaCodecList
import android.media.MediaFormat
import android.view.Surface
import com.tencent.mars.xlog.Log
import kotlin.experimental.and

class MediaCodecSurfaceEncoder {

  companion object {
    private const val TAG = "trinity"
    private const val MIME_TYPE = "video/avc"
    private const val FRAME_INTERVAL = 1
    private const val IDR = 5
    private const val SPS = 7
    private const val TIMEOUT_US: Long = 5000
    private val BYTES_HEADER = byteArrayOf(0, 0, 0, 1)
  }

  private var mEncoder: MediaCodec ?= null
  private var mInputSurface: Surface ?= null
  private val mBufferInfo = MediaCodec.BufferInfo()
  private var mLastPresentationTimeUs: Long = 0

  fun start(width: Int, height: Int, bitRate: Int, frameRate: Int): Int {
    if (width <= 0 || height <= 0 || bitRate <= 0 || frameRate <= 0) {
      Log.e(TAG, "width: $width height: $height, bitRate: $bitRate frameRate: $frameRate")
      return -1
    }
    val format = MediaFormat.createVideoFormat(MIME_TYPE, width, height)
    format.setInteger(MediaFormat.KEY_COLOR_FORMAT, MediaCodecInfo.CodecCapabilities.COLOR_FormatSurface)
    format.setInteger(MediaFormat.KEY_BIT_RATE, bitRate)
    format.setInteger(MediaFormat.KEY_FRAME_RATE, frameRate)
    format.setInteger(MediaFormat.KEY_I_FRAME_INTERVAL, FRAME_INTERVAL)
    try {
      mEncoder = MediaCodec.createEncoderByType(MIME_TYPE)
      mEncoder?.configure(format, null, null, MediaCodec.CONFIGURE_FLAG_ENCODE)
      mInputSurface = mEncoder?.createInputSurface()
      mEncoder?.start()
    } catch (e: Exception) {
      e.printStackTrace()
      Log.e("trinity", e.message)
      return -1
    }
    return 0
  }

  fun getInputSurface(): Surface? {
    return mInputSurface
  }

  fun getLastPresentationTimeUs(): Long {
    return mLastPresentationTimeUs
  }

  fun signalEndOfInputStream() {
    mEncoder?.signalEndOfInputStream()
  }

  fun release() {
    mLastPresentationTimeUs = 0;
    mEncoder?.stop()
    mEncoder?.release()
    mEncoder = null
  }

  private fun isH264StartCode(code: Int): Boolean {
    return code == 0x01
  }

  private fun isH264StartCodePrefix(code: Int): Boolean {
    return code == 0x00
  }

  private fun findAnnexbStartCodecIndex(data: ByteArray, offset: Int): Int {
    var index = offset
    var cursor = offset
    while (cursor < data.size) {
      val code = data[cursor++].toInt()
      if (isH264StartCode(code) && cursor >= 3) {
        val firstPrefixCode = data[cursor - 3].toInt()
        val secondPrefixCode = data[cursor - 2].toInt()
        if (isH264StartCodePrefix(firstPrefixCode) && isH264StartCodePrefix(secondPrefixCode)) {
          break
        }
      }
    }
    index = cursor
    return index
  }

  private fun splitIDRFrame(h264Data: ByteArray): ByteArray {
    var result = h264Data
    var offset = 0
    do {
      offset = findAnnexbStartCodecIndex(h264Data, offset)
      val naluSpecIndex = offset
      if (naluSpecIndex < h264Data.size) {
        val naluType = (h264Data[naluSpecIndex] and 0x1F).toInt()
        if (naluType == IDR) {
          val idrFrameLength = h264Data.size - naluSpecIndex + 4
          result = ByteArray(idrFrameLength)
          System.arraycopy(BYTES_HEADER, 0, result, 0, 4)
          System.arraycopy(h264Data, naluSpecIndex, result, 4, idrFrameLength - 4)
          break
        }
      }
    } while (offset < h264Data.size)
    return result
  }

  fun drainEncoder(data: ByteArray): Int {
    var length = 0
    val encoderOutputBuffers = mEncoder?.outputBuffers ?: return 0
    val status = mEncoder?.dequeueOutputBuffer(mBufferInfo, TIMEOUT_US) ?: return 0
    if (status >= 0) {
      val encodeData = encoderOutputBuffers[status]
      // config data sps/pps
      if ((mBufferInfo.flags and MediaCodec.BUFFER_FLAG_CODEC_CONFIG) != 0) {
        // The codec config_ data was pulled out when we got the
        // INFO_OUTPUT_FORMAT_CHANGED status. The MediaMuxer won't
        // accept
        // a single big blob -- it wants separate csd-0/csd-1 chunks --
        // so simply saving this off won't work.
        if (mBufferInfo.size != 0) {
          encodeData.position(mBufferInfo.offset)
          encodeData.limit(mBufferInfo.offset + mBufferInfo.size)
          length = mBufferInfo.size
          encodeData.get(data, 0, length)
          mBufferInfo.size = 0
        }
      }
      if (mBufferInfo.presentationTimeUs >= mLastPresentationTimeUs) {
        if (mBufferInfo.size != 0) {
          encodeData.position(mBufferInfo.offset)
          encodeData.limit(mBufferInfo.offset + mBufferInfo.size)
          mLastPresentationTimeUs = mBufferInfo.presentationTimeUs
          length = mBufferInfo.size
          encodeData.get(data, 0, mBufferInfo.size)
          val type = (data[4] and 0x1F).toInt()
          if (type == SPS) {
            // 某些手机I帧前面会附带上SPS和PPS，剥离掉SPS和PPS，仅使用I帧
            val idrFrame = splitIDRFrame(data)
            System.arraycopy(idrFrame, 0, data, 0, idrFrame.size)
            length = idrFrame.size
          }
        }
      }
      mEncoder?.releaseOutputBuffer(status, false)
      if ((mBufferInfo.flags and MediaCodec.BUFFER_FLAG_END_OF_STREAM) != 0) {
        // end of stream, break encode data
        return -1
      }
    }
    return length
  }
}